### DISCUSSION TITLE
Wanted: pixel-exact cross-platform icon fonts

### DISCUSSION THREAD

**Discussion by rtollert** [ni/labview-icon-editor#183] [General]
This is something I always wanted to see eventually; I spoke with AristosQueue a couple of years ago and IIRC he opined approximately the same thing.

The Icon Editor relies on LabVIEW's text drawing functions, which in turn relies on platform-specific text rendering behavior. On Windows this is basically guaranteed to do what we want by default, because all fonts are max-hinted at small sizes. But on Linux the fonts tend to come out antialiased. So the same icon operations can result in radically different-looking results, depending on what OS you were making the changes in. Which explains stupid little details like "all of the icon text for the Multiple Errors VIs is antialiased" — the feature was developed mostly on MacOS. (Another problem is that LabVIEW's text compositing is just completely unaware of alpha blending, but AFAIK that's a bug with LabVIEW, not with the Icon Editor.)

Additionally, the Icon Editor uses significantly different default fonts on different platforms. The default font is Small Fonts, because it is far and away the best font for the job on Windows. But that font doesn't exist on Linux, so there, it falls back to Sans. If we were to use a font that exists on all platforms, it would probably need to be Arial, and I'm not particularly thrilled about that.

Potential solutions:
- My gut feeling is that LabVIEW icons, being one of the last vestiges of 32x32 icons in the world, have text rendering requirements which now deviate so far from the rest of the industry that we might be better off designing our own bitmap font, and including a tiny font renderer inside the editor itself. This would also give us the freedom to do things which are otherwise hard to find elsewhere... for instance, we would be able to use an antialiased bitmap font without much trouble if we wanted to, even though that is generally not a thing elsewhere.
- Coax NI into rewriting LabVIEW's text renderer using a cross-platform graphics library. (Note, there are numerous other benefits of such work besides making LabVIEW icon rendering more consistent.) Switch from Small Fonts to something that can be used cross-platform.

**Comment by JayKayAce** [comment]
How would a user go by creating his own font?
Say I wanted to make this font https://youtu.be/wOntiXaIu2o?si=Esz7u35JgFlVlwuT&t=656
![image](https://github.com/ni/labview-icon-editor/assets/24767209/a0fd34c3-d2f0-4c03-b77a-c408645f96cc)
Would that be doable and ship it with the icon editor?

**Comment by rtollert** [comment]
I can't speak as to inclusion, although I will opine that the Icon Editor doesn't have a great track record when it comes to managing user IP — it's always bugged me that I can't define glyph libraries on e.g. a per-project basis. But I haven't thought about that hard enough to distill it into a filable issue.

Character sets are going to be a serious complication. AFAIK, the Icon Editor (like most of the rest of LabVIEW) uses the system encoding. so e.g., in relation to your sample font, LabVIEW literally does not support the use of an encoding which supports literally all of the glyphs in that image. This has serialization implications for the Icon Text tab.

On top of that (and this is something I realized only just now) I think that the statement "writing a bitmap font is pretty easy" is rather Western-centric. Rolling one's own ASCII bitmap font is a thing; rolling one's own CJK bitmap font is, ah, not. And I'm pretty sure that there are many Asian LabVIEW users who would effectively lose major functionality if CJK were unavailable to them in the Icon Editor.

So my first opinion was going to be "just roll our own PNG font atlas format" because traditional bitmap font formats are kind of horrible? But maybe the right approach is PCF. Or, perhaps, both.

**Comment by JayKayAce** [comment]
I think it is an interesting approach to have pr project templates/glyphs, but it would be best to track it in a separate issue ;)

**Comment by svelderrainruiz** [comment]
Will move to a discussion after talking about it on todays steerco meeting

**Comment by jimkring** [comment]
I have drafted a PR that is a start to addressing this need: https://github.com/ni/labview-icon-editor/pull/433
